pkg/userdata/input.go

```diff
@@ -0,0 +1,70 @@
+package userdata
+
+import (
+	"fmt"
+	"strings"
+
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/util/secret"
+)
+
+// ControlPlaneInput defines the context to generate a controlplane instance user data.
+type EtcdPlaneInput struct {
+	BaseUserData
+	secret.Certificates
+	EtcdadmArgs
+
+	EtcdadmInitCommand string
+}
+
+// EtcdPlaneJoinInput defines context to generate etcd instance user data for etcd plane node join.
+type EtcdPlaneJoinInput struct {
+	BaseUserData
+	secret.Certificates
+	EtcdadmArgs
+
+	EtcdadmJoinCommand string
+	JoinAddress        string
+}
+
+// BaseUserData is shared across all the various types of files written to disk.
+type BaseUserData struct {
+	Header              string
+	PreEtcdadmCommands  []string
+	PostEtcdadmCommands []string
+	AdditionalFiles     []bootstrapv1.File
+	WriteFiles          []bootstrapv1.File
+	Users               []bootstrapv1.User
+	NTP                 *bootstrapv1.NTP
+	DiskSetup           *bootstrapv1.DiskSetup
+	Mounts              []bootstrapv1.MountPoints
+	ControlPlane        bool
+	SentinelFileCommand string
+}
+
+type EtcdadmArgs struct {
+	Version         string
+	ImageRepository string
+	EtcdReleaseURL  string
+}
+
+func (args *EtcdadmArgs) SystemdFlags() []string {
+	flags := make([]string, 0, 3)
```

Comment:
I think this should only declare the slice instead of initializing with a length since we are using append as it is, plus the release-url and version flags are option

Reply:
Oh this interesting
This code inits the slice with capacity 3 but length 0
I believe that if I don't allocate memory when I declare it, the go runtime will do it during the first append (and it will possibly allocate way more memory
What do you think?